services/webnn/ort/graph_builder_ort.cc

```diff
@@ -1455,6 +1496,213 @@ void GraphBuilderOrt::AddGemmOperation(const mojom::Gemm& gemm) {
                         attributes);
 }
 
+// `GruType` must be `mojom::Gru` or `mojom::GruCell`.
+template <typename GruType>
+  requires(std::is_same_v<GruType, mojom::Gru> ||
+           std::is_same_v<GruType, mojom::GruCell>)
+[[nodiscard]] base::expected<void, mojom::ErrorPtr>
+GraphBuilderOrt::AddGruOperation(const GruType& gru) {
+  const std::string node_name = GenerateNextOperationName(gru.label);
+
+  uint32_t sequence_lens = 1;
+  uint32_t num_directions = 1;
+  if constexpr (std::is_same_v<GruType, mojom::Gru>) {
+    sequence_lens = gru.steps;
+    num_directions =
+        gru.direction == mojom::RecurrentNetworkDirection::kBoth ? 2 : 1;
+  }
+  const std::vector<uint32_t>& input_shape =
+      GetOperand(gru.input_operand_id).descriptor.shape();
+  CHECK_GE(input_shape.size(), 2u);
+  const uint32_t batch_size = input_shape[input_shape.size() - 2];
+  const uint32_t input_size = input_shape[input_shape.size() - 1];
+  std::string input_name = GetOperandNameById(gru.input_operand_id);
+
+  const uint32_t hidden_size = gru.hidden_size;
+  auto checked_three_times_hidden_size = base::MakeCheckedNum(hidden_size) * 3;
+  CHECK(checked_three_times_hidden_size.IsValid());
+  std::string weight_name = GetOperandNameById(gru.weight_operand_id);
+
+  std::string recurrent_weight_name =
+      GetOperandNameById(gru.recurrent_weight_operand_id);
+
+  if constexpr (std::is_same_v<GruType, mojom::GruCell>) {
+    ASSIGN_OR_RETURN(
+        input_name,
+        PrependReshape(input_name, {sequence_lens, batch_size, input_size}));
+    ASSIGN_OR_RETURN(
+        weight_name,
+        PrependReshape(
+            weight_name,
+            {num_directions, checked_three_times_hidden_size.ValueOrDie(),
+             input_size}));
+    ASSIGN_OR_RETURN(
+        recurrent_weight_name,
+        PrependReshape(
+            recurrent_weight_name,
+            {num_directions, checked_three_times_hidden_size.ValueOrDie(),
+             hidden_size}));
+  }
+
+  std::vector<const char*> input_names = {
+      input_name.c_str(), weight_name.c_str(), recurrent_weight_name.c_str()};
+
+  std::vector<uint32_t> half_bias_dims = {
+      num_directions, checked_three_times_hidden_size.ValueOrDie()};
+  auto half_bias_elements =
+      checked_three_times_hidden_size * num_directions * hidden_size;
+  CHECK(half_bias_elements.IsValid());
+  const OperandDataType input_data_type =
+      GetOperand(gru.input_operand_id).descriptor.data_type();
+  std::string forward_bias_name;
+  if (gru.bias_operand_id.has_value()) {
+    forward_bias_name = GetOperandNameById(gru.bias_operand_id.value());
+    ASSIGN_OR_RETURN(forward_bias_name,
+                     PrependReshape(forward_bias_name, half_bias_dims));
+  } else {
+    switch (input_data_type) {
+      case OperandDataType::kFloat16: {
+        std::vector<uint16_t> forward_bias_value = std::vector<uint16_t>(
+            half_bias_elements.ValueOrDie(), fp16_ieee_from_fp32_value(0.0f));
```

Comment:
To avoid creating tensor with big half_bias_elements, you may want to create a scalar tensor and expand it to target shape.